Auth/Auth.Platform.iOS/Authenticator/OIDCAuthenticator.cs

```diff
@@ -0,0 +1,108 @@
+
+namespace AeroGear.Mobile.Auth.Authenticator
+{
+    public class OIDCAuthenticator : IAuthenticator
+    {
+        private AuthenticationConfig authenticationConfig;
+        private KeycloakConfig keycloakConfig;
+        private ICredentialManager credentialManager;
+
+        private IAuthorizationFlowSession currentAuthorizationFlow;
+
+        private delegate void OIDAuthFlowCallback(OIDCCredential credential, NSError error);
+
+        private TaskCompletionSource<User> authenticateTaskComplete;
+        private Task<User> authenticateTask;
+
+        private IHttpServiceModule httpService;
+
+        public OIDCAuthenticator(AuthenticationConfig authenticationConfig, KeycloakConfig keycloakConfig, ICredentialManager credentialManager, IHttpServiceModule httpServiceModule)
+        {
+            this.authenticationConfig = nonNull(authenticationConfig, "authenticationConfig");
+            this.keycloakConfig = nonNull(keycloakConfig, "keycloakConfig");
+            this.credentialManager = nonNull(credentialManager, "credentialManager");
+            this.httpService = nonNull(httpServiceModule, "httpServiceModule");
+        }
+
+        public Task<User> Authenticate(IAuthenticateOptions authenticateOptions)
+        {
+            IOSAuthenticateOptions options = nonNull((IOSAuthenticateOptions)authenticateOptions, "authenticateOptions");
+            ServiceConfiguration oidServiceConfiguration = new ServiceConfiguration(this.keycloakConfig.AuthenticationEndpoint, this.keycloakConfig.TokenEndpoint); // = GetAuthorizationServiceConfiguration();
+            AuthorizationRequest oidAuthRequest = new AuthorizationRequest(oidServiceConfiguration,
+                                                                           this.keycloakConfig.ResourceId,
+                                                                           new String[] { Scope.OpenId, Scope.Profile },
+                                                                           this.authenticationConfig.RedirectUri,
+                                                                           ResponseType.Code, 
+                                                                           null);
+
+            authenticateTaskComplete = new TaskCompletionSource<User>();
+            authenticateTask = authenticateTaskComplete.Task;
+
+            this.currentAuthorizationFlow = startAuthorizationFlow(oidAuthRequest, options.PresentingViewController, (OIDCCredential credential, NSError error) =>
+            {
+                if (credential != null) {
+                    authenticateTaskComplete.TrySetResult(User.NewUser().FromUnverifiedCredential(credential, this.keycloakConfig.ResourceId));
+                } else {
+                    authenticateTaskComplete.TrySetException(new Exception(error.LocalizedDescription));
+                }
+            });
+
+            return authenticateTask;
+        }
+
+        public async Task<bool> Logout(User currentUser)
```

Comment:
looks like we can create an abstract base class in the Auth.Core module to share some of the code? WDYT?

Reply:
Totally agree